Volatile wild magic

Werner-Dohse opened the pull request:
Some extra functionality to make Wild Surges even more common, is an extension to MORE surges. When Tides of Chaos is used and we have 0/1 resources where the resource label is "Tides of Chaos" (name changed via setting), then also roll a d4 when rolling the d20 for the Wild Surge check. The d4 is used to increase the level cast check, to make Wild Surges more probable. So the check is then d20 <= (spell level cast + d4).

module.zip has update script as well as json file in.
Added my name to authors list if that's not overstepping.

trioderegion commented:
I sincerely appreciate the PR and would love to integrate your work. I need to take some time and see how this fits into the future of 5e helpers and take another pass on WotC copywrite stuff. This feature is stepping further into non-SRD territory and I need to make sure I am handling it correctly.

My current main concern is the increasing number of configuration options. However, I also have plans to utilize tides of chaos in the MORE surges homebrew myself

The following change to your PR would be helpful:

- Combine the Wild Magic homebrew checkboxes into a single dropdown: Standard, More, and Volatile, where Standard = no homebrew, More is my original implementation, and Volatile includes your considerations. Refactoring the surges code may be required. Each wild magic homebrew setting should be as compartmentalized as possible.  You can find code style examples in helpers for handling these dropdowns.

Finally, I'm open to adding more authors, but it does come with responsibilities (such as addressing bugs and implementing planned features). With that said, if you are interested, I'd like to sit down with you and @kandashi to talk about the project and how you can contribute.

Werner-Dohse commented:
Have made the changes as per your request:

- Wild Magic checkboxes (including the Wild Mage Auto Detect checkbox) has been combined into a single dropdown with options: Disabled (default), Enabled - Standard, Enabled - More, Enabled - Volatile. 
- I also refactored the surges code, separated the surges functionality for each type so it's more easily maintainable and reads better. 
- I split the chat notification functionality into it's own function.
- As well as the Wild Surge table roll into it's own function too.
- Added default constant variables for the Feature & Table Names. That way if they select an option but clear those fields, there is a default to fall back to.
- Added console logs for all surge functionality if debug is checked.

The refactor thus makes it easy to remove or add an option from the dropdown if needed.

I'm more than happy to help out where I can, so whenever you want to sit down and talk let me know.

trioderegion commented on scripts/dnd5e-helpers.js:
```diff
@@ -348,32 +359,112 @@ function includes_array(arr, comp) {
+}
+
+/** is the tides of chaos feature used */
+function TidesOfChaosSpent(actor, wmToCFeatureName) {
```
I'm a little confused here regarding the tides of chaos feature. I was assuming the Tides of Chaos feature itself would have a number of uses and we reset the item's usage.

Here it looks like you are assuming the feature uses are stored in a resource?

I think this is the only thing that needs a rework now. Via a call like:
```suggestion
let tcItem = actor.items.getName(wmToCFeatureName);
await tcItem.update( /** reset charges */)
```

TidesOfCahosSpend will need to be async and return the promise given by item.update().

trioderegion commented on module.json:
```diff
@@ -2,7 +2,7 @@
     "title": "DnD5e Helpers",
     "description": "Little helpers for little tasks.",
-    "version": "1.8.4",
+    "version": "1.8.5",
```
Leave this at 1.8.4. I will bump the module version upon official release.

trioderegion commented on module.json:
```diff
@@ -11,6 +11,10 @@
         "name": "Kandashi",
         "url": "https://github.com/kandashi"
+      },
+      {
```
For now, remove this author line. 1.8.5 will be credited to you.

We'll find a time to sit down with you and talk about expanding the dev team 😃

trioderegion reviewed (changes requested):
Great re-work to the PR.

I left comments in the review about the final change I would like to see -- using the feature's charges directly rather than using an actor resource.

After that change, I will merge and release! ❤️

trioderegion commented:
Looks great.  Going to pull in and test!  Give me a day or two for a release :)
